### PR TITLE
Introduce `dropwizard/metrics` integration test

### PR DESCRIPTION
**Pull request by oxkitsune**

### CODE REVIEW

File: integration-tests/metrics-v4.2.19.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19.sh:1]: To ensure the `integration-test.sh` is working for both cases, it'd be nice to also rewrite the content of the `checkstyle-10.12.4.sh` file to sue the script.

File: integration-tests/integration-test.sh [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+set -e -u -o pipefail
+
+integration_test_root="$(readlink -f "$(dirname "${0}")")"
+error_prone_support_root="${integration_test_root}/.."
+repos_root="${integration_test_root}/.repos"
+
+if [ "${#}" -ne 10 ]; then
+  >&2 echo "Usage $(basename "${0}") <test-name> <project> <repository> <revision> <build-flags> <patch-flags> <validation-ep-flags> <validation-mvn-flags> <do-sync> <report-directory>"
+
```
Comment (rickie) [inline integration-tests/integration-test.sh:11]:
```suggestion
```

File: integration-tests/metrics-v4.2.19-expected-changes.patch [hunk 3]
```diff
@@ -0,0 +1,6660 @@
...
+-import java.util.Collections;
+ import java.util.Map;
+ import java.util.Set;
+ import java.util.SortedMap;
+@@ -80,7 +78,7 @@ public class CollectdReporter extends ScheduledReporter {
+     private SecurityLevel securityLevel = SecurityLevel.NONE;
+     private String username = "";
+     private String password = "";
+-    private Set<MetricAttribute> disabledMetricAttributes = Collections.emptySet();
++    private Set<MetricAttribute> disabledMetricAttributes = emptySet();
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19-expected-changes.patch:65]: Interesting, are we not rewriting this to `ImmutableSet.of`?

File: integration-tests/metrics-v4.2.19-expected-warnings.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+metrics-core/src/main/java/com/codahale/metrics/CsvReporter.java:[382,36] [FormatStringConcatenation] Defer string concatenation to the invoked method
+metrics-graphite/src/main/java/com/codahale/metrics/graphite/GraphiteReporter.java:[429,18] [Slf4jLogStatement] Log statement contains 0 placeholders, but specifies 1 matching argument(s)
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19-expected-warnings.txt:2]: Oh nice, this will be interesting to fix upstream! As this is literally a bug!

File: integration-tests/integration-test.sh [hunk 5]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env bash
+
+set -e -u -o pipefail
+
+integration_test_root="$(readlink -f "$(dirname "${0}")")"
+error_prone_support_root="${integration_test_root}/.."
+repos_root="${integration_test_root}/.repos"
+
+if [ "${#}" -ne 10 ]; then
+  >&2 echo "Usage $(basename "${0}") <test-name> <project> <repository> <revision> <build-flags> <patch-flags> <validation-ep-flags> <validation-mvn-flags> <do-sync> <report-directory>"
```
Comment (rickie) [inline integration-tests/integration-test.sh:10]: For this, we should take a look at the other `.sh` files in this repository. I believe we have a slightly different naming situation here (which is closer to the official conventions I believe). 

So I _think_ it should be `[TestName] [Project] ....`

**Review by rickie — APPROVED** [review]
Try to make sure it runs for Java 11 and 17. We discussed this offline :).

Some comments but in general this looks really nice!

**Comment by rickie** [comment]
(Sorry didn't mean to approve yet 😬.)

File: integration-tests/metrics-v4.2.19-expected-changes.patch (hunk 3 above)
Comment (oxkitsune) [reply integration-tests/metrics-v4.2.19-expected-changes.patch:65]: `CollectorMutability` is disabled, because there's a few edge cases iirc. I can supress those edge cases :+1: and re-enable.

File: integration-tests/metrics-v4.2.19-init.patch [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
++        <profile>
++            <id>error-prone-compile</id>
++            <build>
++                <plugins>
++                    <plugin>
++                        <groupId>org.apache.maven.plugins</groupId>
++                        <artifactId>maven-compiler-plugin</artifactId>
++                        <version>${maven-compiler-plugin.version}</version>
++                        <executions>
++                            <!-- Skip the default-compile execution as we don't want to execute the compile goal twice -->
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19-init.patch:76]:
```suggestion
+                            <!-- Skip the `default-compile` execution as we don't want to execute the compile goal twice. -->
```

Same below :).

File: integration-tests/metrics-v4.2.19-init.patch [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                                 <arg>-XDcompilePolicy=simple</arg>
+-                                <arg>-Xplugin:ErrorProne -XepExcludedPaths:.*/target/generated-sources/.*</arg>
+                                 <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
+                                 <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
+                                 <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
+@@ -181,18 +181,129 @@
+                                 <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
+                                 <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
+                             </compilerArgs>
+-                            <annotationProcessorPaths>
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19-init.patch:55]: It doesn't work if we leave this one here combined with the "append" usage in the other profile?

File: integration-tests/metrics-v4.2.19-init.patch [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
++            </build>
++        </profile>
+         <profile>
+             <id>release-sign-artifacts</id>
+             <activation>
+@@ -323,15 +434,7 @@
+                     <compilerArgs>
+                         <arg>-Xlint:all</arg>
+                         <arg>-XDcompilePolicy=simple</arg>
+-                        <arg>-Xplugin:ErrorProne -XepExcludedPaths:.*/target/generated-sources/.*</arg>
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19-init.patch:192]: Do we use this? Otherwise I'd say we can leave this untouched to minimize the diff.

File: integration-tests/integration-test.sh [hunk 9]
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env bash
+# This script is not meant to be invoked manually, instead it should be invoked
```
Comment (rickie) [inline integration-tests/integration-test.sh:2]:
```suggestion

# This script is not meant to be invoked manually, instead it should be invoked
```

Comment (rickie) [reply integration-tests/integration-test.sh:2]: Should we rename to `run-integration-test`? To make clearer what it does.

File: integration-tests/checkstyle-10.12.4.sh [hunk 10]
```diff
@@ -1,11 +1,6 @@
 #!/usr/bin/env bash
-
```
Comment (rickie) [inline integration-tests/checkstyle-10.12.4.sh:2]: Let's not drop this line.

File: .github/workflows/run-integration-tests.yml [hunk 11]
```diff
@@ -32,7 +35,7 @@ jobs:
       - name: Install project to local Maven repository
         run: mvn -T1C install -DskipTests -Dverification.skip
       - name: Run integration test
-        run: xvfb-run ./integration-tests/checkstyle-10.12.4.sh "${{ runner.temp }}/artifacts"
+        run: xvfb-run ./integration-tests/${{ matrix.integration-test }} "${{ runner.temp }}/artifacts"
```
Comment (rickie) [inline .github/workflows/run-integration-tests.yml:38]: There is another mention of `checkstyle-10.12.4` below, so this will not work :).

Comment (rickie) [reply .github/workflows/run-integration-tests.yml:38]: If it's easy we should fix it, but if it's too involved we can defer.

File: integration-tests/integration-test.sh [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+report_directory="${11}"
+
+if [ -n "${report_directory}" ]; then
+  mkdir -p "${report_directory}"
+else
+  report_directory="$(mktemp -d)"
+  trap 'rm -rf -- "${report_directory}"' INT TERM HUP EXIT
+fi
+
+# XXX: Configure Renovate to manage the AssertJ version declared here.
```
Comment (rickie) [inline integration-tests/integration-test.sh:35]: There is no AssertJ version here 😉, it's in the other file.

File: integration-tests/metrics-v4.2.19.sh [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+if [ "${#}" -gt 2 ] || ([ "${#}" = 2 ] && [ "${1:---sync}" != '--sync' ]); then
+  echo "Usage: ${0} [--sync] [<report_directory>]"
+  exit 1
+fi
+do_sync="$([ "${#}" = 0 ] || [ "${1:-}" != '--sync' ] || echo 1)"
+report_directory="$([ "${#}" = 0 ] || ([ -z "${do_sync}" ] && echo "${1}") || ([ "${#}" = 1 ] || echo "${2}"))"
+
+# XXX: We exclude the CollectorMutability and the Immutable*|Preconditions*|StringRules.StringIsNullOrEmpty refaster rules
+# as they introduce changes that expect guava to be on the classpath.
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19.sh:18]:
```suggestion
# as they introduce changes that expect Guava to be on the classpath.
```

File: integration-tests/metrics-v4.2.19.sh [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+revision=v4.2.19
+
+if [ "${#}" -gt 2 ] || ([ "${#}" = 2 ] && [ "${1:---sync}" != '--sync' ]); then
+  echo "Usage: ${0} [--sync] [<report_directory>]"
+  exit 1
+fi
+do_sync="$([ "${#}" = 0 ] || [ "${1:-}" != '--sync' ] || echo 1)"
+report_directory="$([ "${#}" = 0 ] || ([ -z "${do_sync}" ] && echo "${1}") || ([ "${#}" = 1 ] || echo "${2}"))"
+
+# XXX: We exclude the CollectorMutability and the Immutable*|Preconditions*|StringRules.StringIsNullOrEmpty refaster rules
```
Comment (rickie) [inline integration-tests/metrics-v4.2.19.sh:17]:
```suggestion
# XXX: We exclude the `CollectorMutability` and the `Immutable*|Preconditions*|StringRules.StringIsNullOrEmpty` refaster rules
```

Comment (rickie) [reply integration-tests/metrics-v4.2.19.sh:17]: Instead, we also have this flag: https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java#L43 which is used in the CollectorMutability: https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/CollectorMutability.java#L61 😉.

**Review by rickie — COMMENTED** [review]
Some comments 😄.

File: integration-tests/metrics-v4.2.19-init.patch (hunk 7 above)
Comment (oxkitsune) [reply integration-tests/metrics-v4.2.19-init.patch:55]: Leaving this here results in a ` error: plug-in not found: ErrorProne`

File: integration-tests/metrics-v4.2.19-init.patch (hunk 8 above)
Comment (oxkitsune) [reply integration-tests/metrics-v4.2.19-init.patch:192]: Same as above

**Comment by rickie** [comment]
Rebased and resolved conflict.

**Comment by rickie** [comment]
Rebased.

**Comment by oxkitsune** [comment]
Checkstyle needs to be re-run, I will push changes when it's complete.

**Comment by oxkitsune** [comment]
I'm unable to figure out why the refaster options aren't being used in the metrics integration test.
I've tried JDK11 to JDK21, and the exact options given in #654, but I cannot get the `ImmutableSet` rules to disable.

**Comment by rickie** [comment]
Closing in favor of: https://github.com/PicnicSupermarket/error-prone-support/pull/1426.
